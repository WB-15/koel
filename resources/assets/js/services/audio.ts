export const audio = {
  context: null as AudioContext | null,
  source: null as MediaElementAudioSourceNode | null,
  element: null as HTMLMediaElement | null,

  init (element: HTMLMediaElement): void {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext (): AudioContext {
    return this.context!
  },

  getSource (): MediaElementAudioSourceNode {
    return this.source!
  },

  getElement (): HTMLMediaElement {
    return this.element!
  }
}
