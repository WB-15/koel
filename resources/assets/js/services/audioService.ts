export const audioService = {
  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,

  init (element: HTMLMediaElement) {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext () {
    return this.context
  },

  getSource () {
    return this.source
  },

  getElement () {
    return this.element
  }
}
