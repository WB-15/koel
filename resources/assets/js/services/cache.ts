const CACHE_EXPIRATION_TIME = 1000 * 60 * 60 * 24 // 1 day

export const Cache = {
  storage: new Map<any, {
    time: number,
    value: any
  }>(),

  normalizeKey: (key: any) => typeof key === 'object' ? JSON.stringify(key) : key,

  has (key: any) {
    return this.hit(this.normalizeKey(key))
  },

  set (key: any, value: any) {
    this.storage.set(this.normalizeKey(key), {
      value,
      time: Date.now()
    })
  },

  hit (key: any) {
    return !this.miss(this.normalizeKey(key))
  },

  miss (key: any) {
    key = this.normalizeKey(key)

    if (!this.storage.has(key)) return true
    const { time } = this.storage.get(key)!

    if (time < Date.now() - CACHE_EXPIRATION_TIME) {
      this.storage.delete(key)
      return true
    }

    return false
  },

  invalidate (key: any) {
    this.storage.delete(this.normalizeKey(key))
  },

  resolve<T> (key: any, resolver: Closure) {
    key = this.normalizeKey(key)

    this.hit(key) || this.set(key, resolver())
    return this.storage.get(key)!.value
  }
}
