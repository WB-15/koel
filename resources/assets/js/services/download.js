import $ from 'jquery';
import { map } from 'lodash';

import { playlistStore, favoriteStore } from '../stores';
import { ls } from '.';

export const download = {
    /**
     * Download individual song(s).
     *
     * @param {Array.<Object>|Object} songs
     */
    fromSongs(songs) {
        songs = [].concat(songs);
        const ids = map(songs, 'id');
        const params = $.param({ songs: ids });

        return this.trigger(`songs?${params}`);
    },

    /**
     * Download all songs in an album.
     *
     * @param {Object} album
     */
    fromAlbum(album) {
        return this.trigger(`album/${album.id}`);
    },

    /**
     * Download all songs performed by an artist.
     *
     * @param {Object} artist
     */
    fromArtist(artist) {
        // It's safe to assume an artist always has songs.
        // After all, what's an artist without her songs?
        // (See what I did there? Yes, I'm advocating for women's rights).
        return this.trigger(`artist/${artist.id}`);
    },

    /**
     * Download all songs in a playlist.
     *
     * @param {Object} playlist
     */
    fromPlaylist(playlist) {
        if (!playlistStore.getSongs(playlist).length) {
            console.warn('Empty playlist.');
            return;
        }

        return this.trigger(`playlist/${playlist.id}`);
    },

    /**
     * Download all favorite songs.
     */
    fromFavorites() {
        if (!favoriteStore.all.length) {
            console.warn("You don't like any song? Come on, don't be that grumpy.");
            return;
        }

        return this.trigger('favorites');
    },

    /**
     * Build a download link using a segment and trigger it.
     *
     * @param  {string} uri The uri segment, corresponding to the song(s),
     *                      artist, playlist, or album.
     */
    trigger(uri) {
        const sep = uri.indexOf('?') === -1 ? '?' : '&';
        const frameId = `downloader${Date.now()}`;
        $(`<iframe id="${frameId}" style="display:none"></iframe`).appendTo('body');
        document.getElementById(frameId).src = `/api/download/${uri}${sep}jwt-token=${ls.get('jwt-token')}`;
    },
}
