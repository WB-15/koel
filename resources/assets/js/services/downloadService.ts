import { favoriteStore, playlistStore } from '@/stores'
import { authService } from '.'
import { alerts, arrayify } from '@/utils'

export const downloadService = {
  fromSongs (songs: Song | Song[]): void {
    const query = arrayify(songs).reduce((q, song) => `songs[]=${song.id}&${q}`, '')
    this.trigger(`songs?${query}`)
  },

  fromAlbum (album: Album): void {
    this.trigger(`album/${album.id}`)
  },

  fromArtist (artist: Artist): void {
    this.trigger(`artist/${artist.id}`)
  },

  fromPlaylist (playlist: Playlist): void {
    if (playlistStore.getSongs(playlist).length) {
      this.trigger(`playlist/${playlist.id}`)
    }
  },

  fromFavorites (): void {
    if (favoriteStore.all.length) {
      this.trigger('favorites')
    }
  },

  /**
   * Build a download link using a segment and trigger it.
   *
   * @param  {string} uri The uri segment, corresponding to the song(s),
   *                      artist, playlist, or album.
   */
  trigger: (uri: string) => {
    const sep = uri.includes('?') ? '&' : '?'
    const url = `${window.BASE_URL}download/${uri}${sep}api_token=${authService.getToken()}`

    if (KOEL_ENV === 'app') {
      require('electron').ipcRenderer.send('DOWNLOAD', url)
      alerts.success('Download started!')
    } else {
      const iframe = document.createElement('iframe')
      iframe.style.display = 'none'
      iframe.setAttribute('src', url)
      document.body.appendChild(iframe)
    }
  }
}
