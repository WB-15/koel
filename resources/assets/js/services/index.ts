export * from './httpService'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
