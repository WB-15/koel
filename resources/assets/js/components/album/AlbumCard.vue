<template>
  <article
    v-if="showing"
    :class="layout"
    :title="`${album.name} by ${album.artist_name}`"
    class="item"
    data-testid="album-card"
    draggable="true"
    tabindex="0"
    @dblclick="shuffle"
    @dragstart="dragStart"
    @contextmenu.prevent="requestContextMenu"
  >
    <span class="thumbnail-wrapper" :style="{ backgroundImage: `url(${defaultCover})` }">
      <AlbumThumbnail :entity="album"/>
    </span>

    <footer>
      <div class="info">
        <a :href="`#!/album/${album.id}`" class="name" data-testid="name">{{ album.name }}</a>
        <span class="sep text-secondary"> by </span>
        <a v-if="isStandardArtist" :href="`#!/artist/${album.artist_id}`" class="artist">{{ album.artist_name }}</a>
        <span v-else class="text-secondary">{{ album.artist_name }}</span>
      </div>
      <p class="meta">
        <span class="left">
          {{ pluralize(album.song_count, 'song') }}
          •
          {{ duration }}
          •
          {{ pluralize(album.play_count, 'play') }}
        </span>
        <span class="right">
          <a
            :title="`Shuffle all songs in the album ${album.name}`"
            class="shuffle-album"
            data-testid="shuffle-album"
            href
            role="button"
            @click.prevent="shuffle"
          >
            <i class="fa fa-random"/>
          </a>
          <a
            v-if="allowDownload"
            :title="`Download all songs in the album ${album.name}`"
            class="download-album"
            data-testid="download-album"
            href
            role="button"
            @click.prevent="download"
          >
            <i class="fa fa-download"/>
          </a>
        </span>
      </p>
    </footer>
  </article>
</template>

<script lang="ts" setup>
import { computed, defineAsyncComponent, toRef, toRefs } from 'vue'
import { defaultCover, eventBus, pluralize, secondsToHis, startDragging } from '@/utils'
import { albumStore, artistStore, commonStore, songStore } from '@/stores'
import { downloadService, playbackService } from '@/services'

const AlbumThumbnail = defineAsyncComponent(() => import('@/components/ui/AlbumArtistThumbnail.vue'))

const props = withDefaults(defineProps<{ album: Album, layout?: ArtistAlbumCardLayout }>(), { layout: 'full' })
const { album, layout } = toRefs(props)

const allowDownload = toRef(commonStore.state, 'allow_download')

const duration = computed(() => secondsToHis(album.value.length))
const isStandardArtist = computed(() => artistStore.isStandard(album.value.artist_id))
const showing = computed(() => !albumStore.isUnknown(album.value))

const shuffle = async () => {
  await playbackService.queueAndPlay(await songStore.fetchForAlbum(album.value), true /* shuffled */)
}

const download = () => downloadService.fromAlbum(album.value)
const dragStart = (event: DragEvent) => startDragging(event, album.value, 'Album')
const requestContextMenu = (event: MouseEvent) => eventBus.emit('ALBUM_CONTEXT_MENU_REQUESTED', event, album.value)
</script>

<style lang="scss">
.sep {
  display: none;

  .as-list & {
    display: inline;
  }
}

@include artist-album-card();
</style>
