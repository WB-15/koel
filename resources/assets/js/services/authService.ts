import { localStorageService } from '@/services'

const STORAGE_KEY = 'api-token'

export const authService = {
  getToken: () => localStorageService.get<string | null>(STORAGE_KEY),

  hasToken () {
    return Boolean(this.getToken())
  },

  setToken: (token: string) => localStorageService.set(STORAGE_KEY, token),
  destroy: () => localStorageService.remove(STORAGE_KEY)
}
