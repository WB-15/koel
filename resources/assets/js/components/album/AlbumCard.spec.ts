import { fireEvent } from '@testing-library/vue'
import { expect, it } from 'vitest'
import { downloadService, playbackService } from '@/services'
import factory from '@/__tests__/factory'
import UnitTestCase from '@/__tests__/UnitTestCase'
import AlbumCard from './AlbumCard.vue'

let album: Album

new class extends UnitTestCase {
  protected beforeEach () {
    super.beforeEach(() => {
      album = factory<Album>('album', {
        name: 'IV',
        songs: factory<Song>('song', 10)
      })
    })
  }

  protected test () {
    it('renders', () => {
      const { getByText, getByTestId } = this.render(AlbumCard, {
        props: {
          album
        }
      })

      expect(getByTestId('name').textContent).toBe('IV')
      getByText(/^10 songs.+0 plays$/)
      getByTestId('shuffle-album')
      getByTestId('download-album')
    })

    it('downloads', async () => {
      const mock = this.mock(downloadService, 'fromAlbum')

      const { getByTestId } = this.render(AlbumCard, {
        props: {
          album
        }
      })

      await fireEvent.click(getByTestId('download-album'))
      expect(mock).toHaveBeenCalledTimes(1)
    })

    it('shuffles', async () => {
      const mock = this.mock(playbackService, 'playAllInAlbum')

      const { getByTestId } = this.render(AlbumCard, {
        props: {
          album
        }
      })

      await fireEvent.click(getByTestId('shuffle-album'))
      expect(mock).toHaveBeenCalled()
    })
  }
}
