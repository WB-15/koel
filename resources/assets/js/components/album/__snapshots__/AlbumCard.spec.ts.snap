// Vitest Snapshot v1

exports[`renders 1`] = `
<article data-v-f01bdc56="" class="item full" draggable="true" tabindex="0" title="IV by Led Zeppelin"><br data-v-f01bdc56="" data-testid="thumbnail" entity="[object Object]">
  <footer data-v-f01bdc56="">
    <div data-v-f01bdc56="" class="name"><a href="#/album/42" class="text-normal" data-testid="name">IV</a><a href="#/artist/17">Led Zeppelin</a></div>
    <p data-v-f01bdc56="" class="meta"><a title="Shuffle all songs in the album IV" class="shuffle-album" role="button"> Shuffle </a><a title="Download all songs in the album IV" class="download-album" role="button"> Download </a></p>
  </footer>
</article>
`;
